Native profile not compatible with server bootstrapping

I upgraded our spring cloud config server from Finchley.RC1 to RC2 and our integration test broke. The integration test is using a native repository (`spring.profiles.active=native`). Our main bootstrap file has `spring.cloud.config.server.bootstrap=true`

This previously worked fine in RC1. The error I'm getting is in the creation of the DefaultRepositoryConfiguration which should not be created using the native profile from my understanding.

```
Caused by: java.lang.NullPointerException
	at org.springframework.cloud.config.server.environment.HttpClientConfigurableHttpConnectionFactory.addHttpClient(HttpClientConfigurableHttpConnectionFactory.java:60)
	at org.springframework.cloud.config.server.environment.HttpClientConfigurableHttpConnectionFactory.addConfiguration(HttpClientConfigurableHttpConnectionFactory.java:43)
	at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepositoryFactory.build(MultipleJGitEnvironmentRepositoryFactory.java:59)
	at org.springframework.cloud.config.server.config.DefaultRepositoryConfiguration.defaultEnvironmentRepository(EnvironmentRepositoryConfiguration.java:209)
	at org.springframework.cloud.config.server.config.DefaultRepositoryConfiguration$$EnhancerBySpringCGLIB$$b63f2d53.CGLIB$defaultEnvironmentRepository$0(<generated>)
	at org.springframework.cloud.config.server.config.DefaultRepositoryConfiguration$$EnhancerBySpringCGLIB$$b63f2d53$$FastClassBySpringCGLIB$$b07c0778.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:361)
	at org.springframework.cloud.config.server.config.DefaultRepositoryConfiguration$$EnhancerBySpringCGLIB$$b63f2d53.defaultEnvironmentRepository(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 105 more`

When running with --debug flag i get the appropriate configurations based on matching.

`NativeRepositoryConfiguration matched:
      - @ConditionalOnMissingBean (types: org.springframework.cloud.config.server.environment.EnvironmentRepository; SearchStrategy: all) did not find any beans (OnBeanCondition)`

DefaultRepositoryConfiguration:
      Did not match:
         - @ConditionalOnMissingBean (types: org.springframework.cloud.config.server.environment.EnvironmentRepository; SearchStrategy: current) found beans of type 'org.springframework.cloud.config.server.environment.EnvironmentRepository' nativeEnvironmentRepository (OnBeanCondition)
```

My fix is to turn off bootstrapping for the integration test, but I'm sure there are people that would like to use this feature. Thank you so much for this awesome project!

@spencergibb 

I'm sorry, but I don't think this is a duplicate, nor has it been fixed. I have just run with Finchley.RELEASE. The configuration is assuming a git repo when bootstrapping is turned on.

```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'defaultEnvironmentRepository' defined in class path resource [org/springframework/cloud/config/server/config/DefaultRepositoryConfiguration.class]: Invocation of init method failed; nested exception is java.lang.IllegalStateException: You need to configure a uri for the git repository
```

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/diagnostics/GitUriFailureAnalyzer.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/diagnostics/GitUriFailureAnalyzer.java
@@ -1,0 +1,25 @@
+package org.springframework.cloud.config.server.diagnostics;
+
+import org.springframework.boot.diagnostics.AbstractFailureAnalyzer;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.cloud.config.server.environment.JGitEnvironmentRepository;
+
+/**
+ * @author Ryan Baxter
+ */
+public class GitUriFailureAnalyzer extends AbstractFailureAnalyzer<IllegalStateException> {
+
+	public static final String DESCRIPTION = "Invalid config server configuration.";
+	public static final String ACTION = "If you are using the git profile, you need to set a Git URI in your " +
+			"configuration.  If you are using a native profile and have spring.cloud.config.server.bootstrap=true, " +
+			"you need to use a composite configuration.";
+
+	@Override
+	protected FailureAnalysis analyze(Throwable rootFailure, IllegalStateException cause) {
+		if(JGitEnvironmentRepository.MESSAGE.equalsIgnoreCase(cause.getMessage())) {
+			return new FailureAnalysis(DESCRIPTION, ACTION, cause);
+		}
+		return null;
+	}
+
+}

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JGitEnvironmentRepository.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JGitEnvironmentRepository.java
@@ -80,6 +80,8 @@ import static org.eclipse.jgit.transport.ReceiveCommand.Type.DELETE;
 public class JGitEnvironmentRepository extends AbstractScmEnvironmentRepository
 		implements EnvironmentRepository, SearchPathLocator, InitializingBean {
 
+	public static final String MESSAGE = "You need to configure a uri for the git repository.";
+
 	private static final String FILE_URI_PREFIX = "file:";
 
 	private static final String LOCAL_BRANCH_REF_PREFIX = "refs/remotes/origin/";
@@ -243,7 +245,7 @@ public class JGitEnvironmentRepository extends AbstractScmEnvironmentRepository
 	@Override
 	public void afterPropertiesSet() throws Exception {
 		Assert.state(getUri() != null,
-				"You need to configure a uri for the git repository");
+				MESSAGE);
 		initialize();
 		if (this.cloneOnStart) {
 			initClonedRepository();

--- a/spring-cloud-config-server/src/main/resources/META-INF/spring.factories
+++ b/spring-cloud-config-server/src/main/resources/META-INF/spring.factories
@@ -11,3 +11,6 @@ org.springframework.cloud.config.server.bootstrap.ConfigServerBootstrapApplicati
 org.springframework.boot.autoconfigure.EnableAutoConfiguration=\
 org.springframework.cloud.config.server.config.ConfigServerAutoConfiguration,\
 org.springframework.cloud.config.server.config.EncryptionAutoConfiguration
+
+org.springframework.boot.diagnostics.FailureAnalyzer=\
+org.springframework.cloud.config.server.diagnostics.GitUriFailureAnalyzer

--- a/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/NativeBootstrapFailureAnalyzerTests.java
+++ b/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/NativeBootstrapFailureAnalyzerTests.java
@@ -1,0 +1,34 @@
+package org.springframework.cloud.config.server;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.springframework.boot.WebApplicationType;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.boot.test.rule.OutputCapture;
+import org.springframework.cloud.config.server.diagnostics.GitUriFailureAnalyzer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * @author Ryan Baxter
+ */
+public class NativeBootstrapFailureAnalyzerTests {
+
+	@Rule
+	public OutputCapture outputCapture = new OutputCapture();
+
+	@Test
+	public void contextLoads(){
+		try {
+			new SpringApplicationBuilder(ConfigServerApplication.class)
+					.web(WebApplicationType.SERVLET).properties("spring.cloud.bootstrap.name:enable-nativebootstrap").profiles("test","native").run();
+			fail("Application started successfully");
+		}
+		catch (Exception ex) {
+			assertThat(this.outputCapture.toString())
+					.contains(GitUriFailureAnalyzer.ACTION);
+			assertThat(this.outputCapture.toString()).contains(GitUriFailureAnalyzer.DESCRIPTION);
+		}
+	}
+}

--- a/spring-cloud-config-server/src/test/resources/enable-nativebootstrap.yml
+++ b/spring-cloud-config-server/src/test/resources/enable-nativebootstrap.yml
@@ -1,0 +1,5 @@
+spring:
+  cloud:
+    config:
+      server:
+        bootstrap: true

